File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include <bits/stdc++.h>

using namespace std;

int n, s;
int arr[21];
int ans;

void init() {
    cin >> n >> s;
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
}

void dfs(int idx, int sum, int sz) {
    if (idx == n) {
        if (sum == s && sz > 0) {
            ans++;
        }
        return;
    }
    dfs(idx + 1, sum + arr[idx], sz + 1);
    dfs(idx + 1, sum, sz);
}

void solve() {
    dfs(0, 0, 0);
    cout << ans;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);


    init();
    solve();
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.io.*;
import java.util.*;

public class Main{
	static int[] arr;
	static int n;
	static int s;
	static int ans;
	static void init() throws Exception{
		BufferedReader br = new BufferedReader(
				new InputStreamReader(System.in));
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		s = Integer.parseInt(st.nextToken());
		arr = new int[n];
		st = new StringTokenizer(br.readLine());

		for(int i=0;i<n;i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		ans = 0;
		
	}
	static void dfs(int idx, int sum) {
		if(idx>=n) {
			return;
		}
		if(sum+arr[idx]==s) {
			ans++;
		}
		dfs(idx+1, sum+arr[idx]);
		dfs(idx+1,sum);

	}
	public static void main(String[] args) throws Exception {
		init();
		dfs(0,0);
		System.out.println(ans);
	}
}